docs/modules/ROOT/pages/components.adoc

```diff
@@ -0,0 +1,186 @@
+mod MyContract {
+    use openzeppelin::security::InitializableComponent;
+
+    component!(path: InitializableComponent, storage: initializable, event: InitializableEvent);
+
+    // Gives the contract access to the implementation methods
+    impl InitializableImpl =
+        InitializableComponent::InitializableImpl<ContractState>;
+    impl InternalImpl = InitializableComponent::InternalImpl<ContractState>;
+}
+----
+
+Defining the ``impl``s give the contract access to the methods within those implementations from the component.
+For example, `initialize` is defined in the `InternalImpl`.
+It can, therefore, be called like this:
+
+[,javascript]
+----
+#[starknet::contract]
+mod MyContract {
+    (...)
+
+    fn initialize_contract(ref self: ContractState) {
+        self.initializable.initialize();
+    }
+}
+----
+
+TIP: Defining the `InternalImpl` in a contract also gives the contract indirect access to the component's `Storage`.
+
+=== Embedding implementations
+
+A contract can embed implementations into the ABI which will expose the methods of the implementation.
+To embed implementations, add the `#[abi(embed_v0)]` attribute above the `impl`:
+
+[,javascript]
+----
+#[starknet::contract]
+mod MyContract {
+    (...)
+
+    // This attribute exposes the methods of the `impl`
+    #[abi(embed_v0)]
+    impl InitializableImpl =
+        InitializableComponent::InitializableImpl<ContractState>;
+}
+----
+
+`InitializableImpl` defines the `is_initialized` method in the component.
+By adding the embed attribute, `is_initialized` becomes a contract entrypoint for `MyContract`.
+
+=== Component storage and events
+
+The component's storage and events must be added to the contract's `Storage` trait and `Event` enum respectively.
+If the component doesn't define any events, the compiler will still create an empty event enum inside the component module.
+
+[,javascript]
+----
+#[starknet::contract]
+mod MyContract {
+    use openzeppelin::security::InitializableComponent;
+
+    component!(path: InitializableComponent, storage: initializable, event: InitializableEvent);
+
+    (...)
+
+    #[storage]
+    struct Storage {
+        #[substorage(v0)]
+        initializable: InitializableComponent::Storage
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        #[flat]
+        InitializableEvent: InitializableComponent::Event
+    }
+}
+----
+
+The `#[substorage(v0)]` attribute must be included for each component in the `Storage` trait.
+This allows the contract to have indirect access to the component's storage.
+
+The `#[flat]` attribute for events in the `Event` enum, however, is not required.
+Component events are not flattened in the component itself to offer greater flexibility regarding how events are handled.
+Note that if contracts do not flatten component events, the first key in the event log will be the component ID.
+By flattening the component event, the first key will be the event ID.
```

Comment:
Following up our sync, events are not flattened in the component itself because it would remove the event id, not the component scoped id. I think we should rewrite this paragraph.

Reply:
@ericnordelo what do you want to rewrite specifically?